standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -5394,6 +5410,476 @@ public void countOpenTxns() throws MetaException {
+
+        insertIntoPartColStatTable(partitionInfoMap, csId, dbConn);
+
+        for (Map.Entry entry : result.entrySet()) {
+          Map<String, String> parameters = (Map<String, String>) entry.getValue();
+          ColumnStatistics colStats = partColStatsMap.get(entry.getKey());
+          List<String> partVals = getPartValsFromName(tbl, colStats.getStatsDesc().getPartName());
+          if (transactionalListeners != null) {
+            MetaStoreListenerNotifier.notifyEventWithDirectSql(transactionalListeners,
+                    EventMessage.EventType.UPDATE_PARTITION_COLUMN_STAT,
+                    new UpdatePartitionColumnStatEvent(colStats, partVals, parameters,
+                            tbl, writeId, null), dbConn, sqlGenerator);
+          }
+        }
+        dbConn.commit();
+        committed = true;
+        return result;
+      } catch (SQLException e) {
+        checkRetryable(dbConn, e, "updatePartitionColumnStatistics(" + partColStatsMap + "," + tbl + "," + csId
+                + "," + validWriteIds + "," + writeId + ")");
+        throw new MetaException("Unable to update Column stats for  " + tbl.getTableName()
+                + " due to: " + getMessage(e) + "; " + StringUtils.stringifyException(e));
+      } catch (Exception e) {
+        LOG.error("Unable to update Column stats for  " + tbl.getTableName(), e);
+        throw new MetaException("Unable to update Column stats for  " + tbl.getTableName()
+                + " due to: "  + e.getMessage());
+      } finally {
+        if (!committed) {
+          rollbackDBConn(dbConn);
+        }
+        closeDbConn(dbConn);
+        unlockInternal();
+      }
+    } catch (RetryException e) {
+      return updatePartitionColumnStatistics(partColStatsMap, tbl, csId, validWriteIds, writeId);
+    }
+  }
+
+  @Override
+  public long getNextCSIdForMPartitionColumnStatistics(long numStats) throws MetaException {
+    Statement statement = null;
+    ResultSet rs = null;
+    long maxCsId = 0;
+    boolean committed = false;
+    Connection dbConn = null;
+
+    try {
+      lockInternal();
+      dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED, connPoolMutex);
+
+      // This loop will be iterated at max twice. If there is no records, it will first insert and then do a select.
+      while (maxCsId == 0) {
+        String query = "SELECT \"NEXT_VAL\" FROM \"SEQUENCE_TABLE\" WHERE \"SEQUENCE_NAME\"= "
+                + quoteString("org.apache.hadoop.hive.metastore.model.MPartitionColumnStatistics")
+                + " FOR UPDATE";
+        LOG.debug("Going to execute query " + query);
+        statement = dbConn.createStatement();
+        rs = statement.executeQuery(query);
+        if (rs.next()) {
+          maxCsId = rs.getLong(1);
+        } else {
+          closeStmt(statement);
+          statement = dbConn.createStatement();
+          query = "INSERT INTO \"SEQUENCE_TABLE\" (\"SEQUENCE_NAME\", \"NEXT_VAL\")  VALUES ( "
```

Comment:
shouldn't you insert the next value and return it right away?

Reply:
It may possible that some other txn  has already inserted and we just have to select. I will add a catch to check for duplicate records during insert.

Reply:
done